src/cs/Ssh/SecureStream.cs

```diff
@@ -0,0 +1,291 @@
+		this.session.Closed += OnSessionClosed;
+
+		this.clientCredentials = clientCredentials ??
+			throw new ArgumentNullException(nameof(clientCredentials));
+	}
+
+	/// <summary>
+	/// Creates a secure stream over an underlying transport stream, using server credentials
+	/// to authenticate.
+	/// </summary>
+	/// <param name="transportStream">Underlying (insecure) transport stream.</param>
+	/// <param name="serverCredentials">Server authentication credentials.</param>
+	/// <param name="trace">Optional trace source for SSH protocol tracing.</param>
+	public SecureStream(
+		Stream transportStream,
+		SshServerCredentials serverCredentials,
+		TraceSource? trace = null)
+	{
+		this.transportStream = transportStream ??
+			throw new ArgumentNullException(nameof(transportStream));
+		this.session = new SshServerSession(
+			SshSessionConfiguration.Default,
+			trace ?? new TraceSource(nameof(MultiChannelStream)));
+
+		this.session.Authenticating += OnSessionAuthenticating;
+		this.session.Closed += OnSessionClosed;
+
+		this.serverCredentials = serverCredentials ??
+			throw new ArgumentNullException(nameof(serverCredentials));
+	}
+
+	public event EventHandler<SshAuthenticatingEventArgs>? Authenticating;
+
+	/// <summary>
+	/// Gets a value indicating whether the session is closed.
+	/// </summary>
+	public bool IsClosed =>
+		this.session.IsClosed;
+
+	/// <summary>
+	/// Event that is rised when underlying ssh session is closed.
+	/// </summary>
+	/// <remarks>
+	/// The event is rised before the session stream is closed.
+	/// The stream will be closed after the event handler.
+	/// </remarks>
+	public EventHandler<SshSessionClosedEventArgs>? Closed
+	{
+		get;
+		set;
+	}
+
+	/// <summary>
+	/// Limits the amount of time that ConnectAsync() may wait for the initial
+	/// session handshake (version exchange).
+	/// </summary>
+	public TimeSpan? ConnectTimeout
+	{
+		get => this.session.ConnectTimeout;
+		set => this.session.ConnectTimeout = value;
+	}
+
+	/// <summary>
+	/// Initiates the SSH session over the transport stream by exchanging initial messages with the
+	/// remote peer. Waits for the protocol version exchange, key exchange, authentication, and
+	/// opened channel. Additional message processing is kicked off as a background task chain.
+	/// </summary>
+	/// <exception cref="SshConnectionException">The connection failed due to a protocol
+	/// error.</exception>
+	/// <exception cref="TimeoutException">The ConnectTimeout property is set and the initial
+	/// version exchange could not be completed within the timeout.</exception>
+	public async Task ConnectAsync(CancellationToken cancellation = default)
+	{
+		if (Authenticating == null)
+		{
+			throw new InvalidOperationException(
+				"An Authenticating event handler must be registered before connecting.");
+		}
+
+		if (this.serverCredentials != null)
+		{
+			var serverSession = (SshServerSession)this.session;
+			serverSession.Credentials = this.serverCredentials;
```

Comment:
Why do we assign server credentials here and not in ctor?

Reply:
It could be done in either place. But this method is setting up both server and client authentication and is handling client credentials (which cannot be set in the constructor), so I thought it is easier to understand if this method handled the server credentials also.